Improve Error Checking Code

viceroypenguin opened the pull request:
This PR migrates code contracts for methods to `CommunityToolkit.Diagnostics` support methods. This has several advantages:
* Improves chance that error-checking host method will be inlined, possibly improving performance
* Reduces possibility human error by specifying parameter name only once
* Reduces boilerplate code.

Fixes #903 

Tasks:
* [x] `ArgumentNullException`
* [x] `ArgumentOutOfRangeException`
* [x] `ArgumentException`
* [x] `InvalidOperationException` (??)

atifaziz reviewed (changes requested):
Thanks for picking up #903 and your effort here!

I noticed one sweeping so could I start by asking you to fix misalignments due to tabs? You can find these with:

    git grep -P '\t' -- *.cs *.tt *.csproj

I'll get round to a more detailed review after. Thanks!

atifaziz commented on MoreLinq/Fold.cs:
```diff
@@ -31,14 +31,14 @@ static T[] Fold<T>(this IEnumerable<T> source, int count)
                 elements[i] = i < count ? item : throw LengthError(1);
                 i++;
-            }
+        }
```
Revert:

```suggestion
            }
```

atifaziz commented on MoreLinq/Fold.cs:
```diff
@@ -31,14 +31,14 @@ static T[] Fold<T>(this IEnumerable<T> source, int count)
-        }
     }
 }
+}
```
Something seems to have gone off with the indentation here.

atifaziz commented on MoreLinq/Fold.g.tt:
```diff
@@ -32,6 +32,7 @@
 
 namespace MoreLinq
 {
+    using CommunityToolkit.Diagnostics;
```
Remove unused import:

```suggestion
```

atifaziz commented on MoreLinq/Fold.g.cs:
```diff
@@ -27,6 +27,7 @@
 
 namespace MoreLinq
 {
+    using CommunityToolkit.Diagnostics;
```
Remove unused import:

```suggestion
```

atifaziz commented on MoreLinq/Append.cs:
```diff
@@ -17,7 +17,7 @@
 namespace MoreLinq
 {
-    using System;
+	using CommunityToolkit.Diagnostics;
```
Convert tab to 4 spaces:

```suggestion
    using CommunityToolkit.Diagnostics;
```

atifaziz reviewed (commented):
Thanks for the follow-up. There are naturally tons of changed files so it's going to take me a few sittings, but I managed get through a quarter of the 120 files and so here are some interim/initial thoughts…

I noticed that we're only using a tiny fraction of the guards from the **[CommunityToolkit]**

- `Guard.IsBetween`
- `Guard.IsBetweenOrEqualTo`
- `Guard.IsGreaterThan`
- `Guard.IsGreaterThanOrEqualTo`
- `Guard.IsLessThan`
- `Guard.IsLessThanOrEqualTo`
- `Guard.IsNotNull`

I'm therefore leaning towards it being simpler to just embed these into the project instead of taking an additional dependency. A downside might be eventual divergence with time, but I don't expect these to change much and it would also enable some flexibility for tailoring. For example, I wonder if [the `Guard` methods shouldn't be changed to _return_ the valid argument, for some future-readiness](https://twitter.com/raboof/status/1683944201275473921).

Do we really need a throw helper for `InvalidOperationException`? I think you had doubts about the value of this too? I don't think it helps the JIT case. Consistency might be one argument, but a bare `throw` still reads simpler. I can understand doing this consistently for arguments, but invalid use seems like an outlier.

[CommunityToolkit]: https://www.nuget.org/packages/CommunityToolkit.Diagnostics/8.2.2

viceroypenguin commented:
> I'm therefore leaning towards it being simpler to just embed these into the project instead of taking an additional dependency. A downside might be eventual divergence with time, but I don't expect these to change much and it would also enable some flexibility for tailoring. 

If you want to take the time to copy them from CommunityToolkit and implement them yourself, I won't disagree. That said, a) I'm far too lazy to redo what someone else has already done, and b) I think this goes against principles of working with libraries in the first place.

Namely if _every_ library copies these methods, then now these methods are implemented in each library, increasing dll size and JIT size since they all have to be JIT'd separately, especially for generic methods which need to be genericized separately for each implementation. 

Using a single library (CommunityToolkit for cross-TFM, BCL for net8+ which has many of these now implemented) means that this code is shared across all consumers. 

> For example, I wonder if [the Guard methods shouldn't be changed to return the valid argument, for some future-readiness](https://twitter.com/raboof/status/1683944201275473921).

I used to think that too, when I started with the Throw library (from amantinband) and switched to CommunityToolkit. However, I've found as I've used `Guard.Xxx` more, I now prefer the non-return version because it encourages better code readability, specifically, that contractual obligations (not-null, not-null/empty/whitespace, numeric range) are specified and checked at the start of the method.

> Do we really need a throw helper for `InvalidOperationException`? I think you had doubts about the value of this too? 

The advantage of this specifically is two-fold:
1. the calling code reduces JIT code as the `throw` creates a fairly large footprint
2. calling code is no-longer allowed to be inlined (inlining is turned off for any method that includes a `throw` anywhere in the method). This may not be as big of a deal for a `throw` in the middle of an iterator method, but definitely has an impact on methods that simply return a value. Obviously all of the outer methods that MoreLinq uses are more likely to be inlined, switching to `Guard.Xxx`, which means many of them may turn into direct calls to the `Core` method if the JIT can prove that the values are not-null, etc.
   
   Whether this matters in the cases of the various `ThrowHelper` calls is less certain, but definitely `TryAsCollectionLike()` and `AggregateRight()` will benefit from it. For me, I've grown used to it, so I now actually prefer `ThrowHelper.` instead of `throw new`, but I'll concede that to be a usage/personal preference.

atifaziz reviewed (changes requested):
I did another review session and posted some comments. As I did the review, I had some more overall thoughts about this.

We're only using a handful of `Guard` and `ThrowHelper` methods from the CommunityToolkit and most of them are plain one-liners. What's more, practically all of the ones we need have found their way into .NET 8. Some of the complexity that comes from the [friendly type string formatting](https://github.com/CommunityToolkit/dotnet/blob/v8.2.2/src/CommunityToolkit.Diagnostics/Extensions/TypeExtensions.cs) that CommunityToolkit does in the error messages doesn't seem to be warranted. This is not about redoing or not leveraging a library, but assessing whether taking a whole dependency is worth 7 dead simple methods that are landing in .NET anyway. The considerations and criteria are often different for an application than an F/OSS library.

I think we should leverage the methods added in .NET instead of relying on the CommunityToolkit. For targets where the helpers are unavailable, we can copy over the 1-2 line implementations from .NET and light them up conditionally. These will only need to be maintained until those targets have been dropped, although I don't expect there'll any maintenance to do since they can't be expected to change much at all with time.

In related issue, you proposed two options:

> Personal recommendation: add a reference to [`CommunityToolkit.Diagnostics`](https://www.nuget.org/packages/CommunityToolkit.Diagnostics/), and update all guard statements to these. This would allow us to rely on a shared implementation of this boilerplate code and if others do as well, then that is mutual code shared.
>
> Alternative recommendation: Add our own `Guard` and `ThrowHelper` methods and update all guard statements to these. This would duplicate code that could be shared, but not add a dependency on another library.

I'm still leaning on the second one, but saying that let's go with what's there & coming with .NET.

---
PS The whole business of helping JIT inline the argument validation is interesting for methods that are expected to be used on hot paths, but since that's where one generally avoids LINQ anyway, the overall benefits are marginal. Moreover, tiered compilation and PGO may never inline in the first place so one may not even be getting the benefits one might expect!

viceroypenguin commented on MoreLinq.Test/Extensions.cs:
```diff
@@ -17,6 +17,7 @@
 
 namespace MoreLinq.Test
 {
+    using CommunityToolkit.Diagnostics;
```
auto-add code-fix. usually it respects that, but alas...

viceroypenguin commented:
> For targets where the helpers are unavailable, we can copy over the 1-2 line implementations from .NET and light them up conditionally. These will only need to be maintained until those targets have been dropped, although I don't expect there'll any maintenance to do since they can't be expected to change much at all with time.

This is not a bad suggestion. I'll have to think about how to implement in a way that a) isn't too complicated and b) doesn't require tfm based evaluations at the call site. 

What are your thoughts about using global usings to hide the complexity. Could do something like:
```xml
<ItemGroup Condition="$(TargetFramework) != 'net8.0'">
  <Using Include="MoreLinq.Exceptions.ArgumentNullException" Static="true"/>
</ItemGroup>
<ItemGroup Condition="$(TargetFramework) == 'net8.0'">
  <Using Include="System.ArgumentNullException" Static="true"/>
</ItemGroup>
```
```cs
public static IEnumerable<TSource> Operator<TSource>(this IEnumerable<TSource> source)
{
  ThrowIfNull(source);
  // ...
}
```

atifaziz commented:
> What are your thoughts about using global usings to hide the complexity.

@viceroypenguin I was thinking something along the very same lines! However, instead of statically importing methods, I'd like to try and maintain the type name qualification. Taking the example of one of the `Pad` overloads, ideally the code looks as you'd write it in .NET 8:

```c#
public static IEnumerable<TSource> Pad<TSource>(this IEnumerable<TSource> source, int width,
                                                Func<int, TSource> paddingSelector)
{
    ArgumentNullException.ThrowIfNull(source);
    ArgumentNullException.ThrowIfNull(paddingSelector);
    ArgumentOutOfRangeException.ThrowIfLessThan(width, 0);

    return PadImpl(source, width, default, paddingSelector);
}
```

The targets that won't have those methods on types could have masking definitions in the `MoreLinq` namespace (which could be made to take precedence), like so:

```c#
#if EXCEPTION_HELPERS

namespace MoreLinq
{
    using System.Diagnostics.CodeAnalysis;
    using System.Runtime.CompilerServices;

    static class ArgumentNullException
    {
        public static void ThrowIfNull([NotNull] object? argument,
                                       [CallerArgumentExpression(nameof(argument))]
                                       string? paramName = null)
        {
            if (argument is null)
                Throw(paramName);
        }

        [DoesNotReturn]
        static void Throw(string? paramName) =>
            throw new System.ArgumentNullException(paramName);
    }
}

#endif
```

A conditional symbol like `EXCEPTION_HELPERS` could then be defined in the project file based on the target framework. This is just a rough idea from the head and a little bit of experimentation may be needed to find the sweet spot.

This does mean however that you can't just write `throw new ArgumentNullException()` anymore, but that may not be a bad thing!

I would also limit this just to the main library project. I don't think we need to extend this approach, for example, to the test project.

What would be the cherry on top is to add an integration test that exercises the inlining (although as I said previously, tiered compilation means you don't always get the expected optimisations). I have some ideas for how this could be pulled off.

atifaziz commented:
@viceroypenguin As a pulse check, do you plan to follow-up here based on latest discussion? I'm just asking to triage open issues and PRs.

viceroypenguin commented:
Yes, but will probably be a while. That's a bit of more work than I have time for right now.

viceroypenguin commented:
@atifaziz Question on which path you want to go here. 

Currently, the following code does not work:
```cs
#if !NET6_0_OR_GREATER
global using ArgumentNullException = MoreLinq.Exceptions.ArgumentNullException;
#endif
```

The reason is that each `namespace MoreLinq { using System; }` creates a tighter binding for `ArgumentNullException` to `System` than to the global alias. 

I don't have this problem in SuperLinq, because I migrated to implicit usings and file-scoped namespaces. Since these are the recommended patterns, I'd recommend we do a PR to do these migrations. 

Alternatively, each file can include 
```cs
#if !NET6_0_OR_GREATER
global using ArgumentNullException = MoreLinq.Exceptions.ArgumentNullException;
#endif
```
at the top of the file, below the `using System;`. I think this path is a huge detriment to the usability of each file.

Your call, though.
